synapseclient/core/utils.py

```diff
@@ -339,15 +339,58 @@ def is_same_base_url(url1: str, url2: str) -> bool:
     return url1.scheme == url2.scheme and url1.hostname == url2.hostname
 
 
-def is_synapse_id_str(obj):
+def is_synapse_id_str(obj: str) -> typing.Union[str, None]:
     """If the input is a Synapse ID return it, otherwise return None"""
     if isinstance(obj, str):
-        m = re.match(r"(syn\d+$)", obj)
+        m = re.match(r"(syn\d+(\.\d+)?$)", obj)
         if m:
             return m.group(1)
     return None
 
 
+def get_synid_and_version(
+    obj: typing.Union[str, collections.abc.Mapping]
+) -> typing.Tuple[str, typing.Union[int, None]]:
+    """Extract the Synapse ID and version number from input entity
+
+    Arguments:
+            obj: May be a string, Entity object, or dictionary.
+
+    Returns:
+        A tuple containing the synapse ID and version number,
+        where the version number may be an integer or None if
+        the input object does not contain a versonNumber or
+        .version notation (if string).
+
+    Example: Get synID and version from string object
+        Extract the synID and version number of the entity string ID
+
+            from synapseclient.core import utils
+            utils.get_synid_and_version("syn123.4")
+
+        The call above will return the following tuple:
+
+            ('syn123', 4)
+    """
+
+    if isinstance(obj, str):
+        synapse_id_and_version = is_synapse_id_str(obj)
+        if not synapse_id_and_version:
+            raise ValueError("The input string was not determined to be a syn ID.")
+        m = re.match(r"(syn\d+)(?:\.(\d+))?", synapse_id_and_version)
+        id = m.group(1)
+        version = int(m.group(2)) if m.group(2) is not None else m.group(2)
+
+        return id, version
+
+    id = id_of(obj)
+    version = None
+    if "versionNumber" in obj:
+        version = obj["versionNumber"]
+
+    return id, version
```

Comment:
Nit: Just a thought for the future, but tuples can be a bit risky because the order really matters.  Look into `namedtuples` for something more explicit.